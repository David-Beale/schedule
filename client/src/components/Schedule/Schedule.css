.outer-container{
  max-height:100vh;
  overflow: hidden;
}

.container{
  width:100vw;
  position: relative;
  z-index: 0;
}
.flex-container{
  display:flex;
}
.c1{
  height:3vh;
  margin-top:5vh;
  overflow-y: visible;
  position: relative;
  z-index: 5;
}
.fc1{
  width:100vw;
  height:3vh;
  overflow-x:hidden;
  overflow-y: visible;
}

.c2{
  height:92vh;
  margin-top:0vh;
  overflow:scroll;
  position: relative;
  z-index: 0;
}

.fc2{
  height:200px;
}
.grid-item1 {
  box-sizing: border-box;
  background-image: linear-gradient(rgba(63, 63, 63, 0.8), rgba(139, 137, 137, 0.8) ) ;
  height:100%;
  color: aliceblue;
  border: 1px solid rgba(0, 0, 0, 0.8);
  font-size: 18px;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  padding-left: 20px;
  min-width:175px;
}
.end{
  min-width:175px
}
.grid-item2 {
  box-sizing: border-box;
  height: 200px;
  min-width: 175px;
  border: 1px solid rgba(122, 122, 122, 0.322);
  font-size: 30px;
  display: flex;
  align-items: center;
  justify-content: center;
  color: rgb(3, 135, 250);
  position: relative;
}

.now{
  width:5px;
  background-color: rgba(114, 139, 247, 0.562);
  position: absolute;
  top:0;
  z-index: 11;
}
.content{
  box-sizing: border-box;
  height:97.5%;
  margin-top: 2.5%;
  background: linear-gradient(to bottom, #f7f7f7 0%,#e5eaec 70%,#e7f0f7 100%); 
  position: absolute;
  top:0;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  flex-direction: column;
  z-index: 10;
  border: 1px solid rgb(128, 127, 127);
  color: rgb(63, 63, 63);
  overflow: scroll;
}
.content:hover{
  box-shadow: 0px 2px 6px 1px rgba(0,0,0,0.75);
  background: linear-gradient(to bottom, #f7f7f7 0%,#dbeff7 50%,#d1e4f3 100%); 
}
.content-title{
  font-family: 'Palanquin', sans-serif;
  color: #f05537;
  font-size: 1.6rem;;
  font-weight: 700;
  max-height: 2rem;
  overflow: scroll;
}
.content-image-container{
  width:175px;
  height:50%;
  overflow: hidden;
}
.content-image{
  width:100%;
}
.content-description{
  font-family: 'Palanquin', sans-serif;
  color: #000000;
  font-size: 1.4rem;
  font-weight: 400;
  max-height: 4rem;
  overflow: scroll;
}
/* width */
::-webkit-scrollbar {
  display: none;
}
.c2::-webkit-scrollbar {
  display: initial;
  width: 10px;
  height: 10px;
}

/* Track */
::-webkit-scrollbar-track {
  background: #f1f1f1;
}

/* Handle */
::-webkit-scrollbar-thumb {
  background: #888;
  border-radius: 5px;
}

/* Handle on hover */
::-webkit-scrollbar-thumb:hover {
  background: #555;
}