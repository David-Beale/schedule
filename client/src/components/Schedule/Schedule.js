import React, { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import Moment from 'react-moment';
import ScheduleCard from '../Schedule-card'

function Schedule (props) {
  const [rowCounter, setRowCounter] = useState(1);
  const [rows, setRows] = useState([]);
  const [initialRow, setInitialRow] = useState([]);
  const [hoursArray, setHoursArray] = useState([]);
  const [data, setData] = useState([]);
  const [filteredData, setFilteredData] = useState([]);
  const dataStart = useSelector(({ eventsReducer }) => eventsReducer.events);
  const today = new Date();
  const [currentDay, setCurrentDay] = useState(today);
  const [positions, setPositions] = useState([])
  const [renderCards, setRenderCards] = useState(false)
  const hours = today.getHours();
  const minutes = today.getMinutes();


  //import data store
  useEffect(() => {
    if (dataStart.length > 0) {
      let array = dataStart.map((event) => event.eventData);
      setData(array);
      filterByDay(array, today);
    }
  }, [dataStart]);
  useEffect(() => {
    //setup time top margin
    let localHours = [];
    for (let i = 0; i < 24; i++) {
      let index = i.toString();
      if (index.length < 2) index = '0' + index;
      if (i < 23)
        localHours.push(
          <div
            id={`margin${i}`}
            className="grid-item1"
            key={i}
          >{`${index}:00`}</div>
        );
      else
        localHours.push(
          <div className="grid-item1 end" key={i}>{`${index}:00`}</div>
        );
    }
    setHoursArray(localHours);
    //setup first row
    const tempRows = [];
    const content = [];
    for (let i = 0; i < 24; i++) {
      content.push(<div className="grid-item2" id={`time${i}`} key={i}></div>);
    }
    tempRows.push(
      <div className="flex-container fc2" id={`row${0}`} key={0}>
        {content}
      </div>
    );
    setInitialRow(tempRows);
    setRows(tempRows);
  }, []);

  const unitWidth = 175;

  useEffect(() => {
    if (hoursArray.length > 0) {
      const slider = document.querySelector('.c2');
      const slider2 = document.querySelector('.fc1');

      //setup current time indicator
      let time = unitWidth * hours + unitWidth * (minutes / 60);
      slider.scrollLeft = time - window.innerWidth / 2;
      const timeDiv = document.createElement('div');
      timeDiv.className = 'now';
      timeDiv.style.left = `${(minutes / 60) * 100}%`;
      timeDiv.style.height = `${200 * rowCounter}px`;
      document.querySelector(`#time${hours}`).appendChild(timeDiv);

      //setup drag to scroll
      let isDown = false;
      let startX;
      let startY;
      let scrollLeft;
      let scrollTop;
      let dragging = false;

      slider.addEventListener('mouseenter', (e) => {
        e.preventDefault();
      });
      slider.addEventListener('mousedown', (e) => {
        isDown = true;
        startX = e.pageX - slider.offsetLeft;
        startY = e.pageY - slider.offsetTop;
        scrollLeft = slider.scrollLeft;
        scrollTop = slider.scrollTop;
      });
      slider.addEventListener('scroll', (e) => {
        if (dragging) return;
        slider2.scrollLeft = slider.scrollLeft;
      });
      slider.addEventListener('mouseup', () => {
        isDown = false;
        dragging = false;
      });
      slider.addEventListener('mousemove', (e) => {
        if (!isDown) return;
        e.preventDefault();
        dragging = true;
        const x = e.pageX - slider.offsetLeft;
        const y = e.pageY - slider.offsetTop;
        const walkx = x - startX;
        const walky = y - startY;
        slider.scrollLeft = scrollLeft - walkx;
        slider.scrollTop = scrollTop - walky;
        slider2.scrollLeft = slider.scrollLeft;
      });
    }
  }, [hoursArray]);
  useEffect(() => {
    // Database info processing
    if (hoursArray.length > 0 && filteredData.length > 0) {
      const tempRows = initialRow.slice();
      let localRowCounter = 1;
      let localRowStorage = [Array(96)];
      let localPositions = []
      for (let i = 0; i < filteredData.length; i++) {
        const contentHours = +filteredData[i].date.getHours();
        const contentMins = +filteredData[i].date.getMinutes();
        let contentLength =
          (filteredData[i].endTime.getTime() - filteredData[i].date.getTime()) /
          1000 /
          60;
        if (contentLength === 0) continue;
        const startBlock = Math.floor(contentHours * 4 + contentMins / 15);
        let blockEnd = Math.ceil(
          contentHours * 4 + contentMins / 15 + contentLength / 15
        );
        if (blockEnd > 96) {
          blockEnd = 96;
        }
        let spaceCheck;
        let selectedRow;
        for (let i = 0; i < localRowStorage.length; i++) {
          spaceCheck = true;
          for (let j = startBlock; j < blockEnd; j++) {
            if (localRowStorage[i][j] === 1) {
              spaceCheck = false;
              j = blockEnd;
            } else {
              localRowStorage[i][j] = 1;
            }
          }
          if (spaceCheck) {
            selectedRow = i;
            i = localRowStorage.length;
          }
        }
        if (!spaceCheck) {
          localRowStorage.push(Array(96));
          for (let j = startBlock; j < blockEnd; j++) {
            localRowStorage[localRowStorage.length - 1][j] = 1;
          }
          const content = [];
          for (let i = 0; i < 24; i++) {
            content.push(<div className="grid-item2" id={`time${i}`} key={i}></div>);
          }
          tempRows.push(
            <div className="flex-container fc2" id={`row${localRowCounter}`} key={localRowCounter}>
              {content}
            </div>
          );
          localRowCounter++;
          selectedRow = localRowCounter - 1;
        }
        localPositions.push({ selectedRow, startBlock, blockEnd })
      }
      setRowCounter(localRowCounter);
      setPositions(localPositions)
      setRenderCards(true)
      setRows(tempRows);
    } else if (hoursArray.length > 0 && filteredData.length === 0) {
      setRows(initialRow);
      setRowCounter(1);
    }
  }, [filteredData]);
  useEffect(() => {
    const timeDiv = document.querySelector('.now');
    if (timeDiv) timeDiv.style.height = `${200 * rowCounter}px`;
  }, [rowCounter]);

  const dayBack = () => {
    let day = currentDay;
    day = new Date(day.setDate(day.getDate() - 1));
    setCurrentDay(day);
    setRenderCards(false)
    filterByDay(data, day);
  };
  const dayFwd = () => {
    let day = currentDay;
    day = new Date(day.setDate(day.getDate() + 1));
    setCurrentDay(day);
    setRenderCards(false)
    filterByDay(data, day);
  };
  const filterByDay = (array, filterDay) => {
    let year = filterDay.getFullYear();
    let month = filterDay.getMonth();
    let day = filterDay.getDate();
    let now = new Date(year, month, day);
    let filtered = array.filter((event) => {
      let eventDate = event.date;
      let evetYear = eventDate.getFullYear();
      let eventMonth = eventDate.getMonth();
      let eventDay = eventDate.getDate();
      eventDate = new Date(evetYear, eventMonth, eventDay);
      return eventDate - now === 0;
    });
    setFilteredData(filtered);
  };
  return (
    <div className="outer-container">
      <div className="container date">
        <div onClick={dayBack} className="arrow">&#x2B05;</div>{' '}<Moment className="sched-time" format="ddd Do MMM">{currentDay}</Moment>{' '}<div onClick={dayFwd} className="arrow">&#x27A1;
        </div>
      </div>
      <div className="container c1">
        <div className="flex-container fc1">{hoursArray}</div>
      </div>

      <div className="container c2">
        {rows}
        {renderCards && filteredData.map((event, index) => {
          return (
            <ScheduleCard
              key={index}
              event={event}
              unitWidth={unitWidth}
              hours={hours}
              minutes={minutes}
              position={positions[index]}
            />
          )
        })}
      </div>
    </div>
  );
}

export { Schedule as default };
