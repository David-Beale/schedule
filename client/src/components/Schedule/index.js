import Schedule from './Schedule';
import './schedule.css';

export { Schedule as default };
