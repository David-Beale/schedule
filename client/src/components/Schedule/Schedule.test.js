import Schedule from './Schedule';
import { render } from '@testing-library/react';
import ReactDOM from 'react-dom';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from '../../redux/store';

const store = configureStore();

let windowSpy;

describe('Schedule', () => {
  const RealDate = Date.now;

  beforeAll(() => {
    global.Date.now = jest.fn(() => new Date('2019-04-07T10:20:30Z'));
  });

  afterAll(() => {
    global.Date.now = RealDate;
  });
  it('should render properly', () => {
    const today = jest.fn(() => new Date('2019-04-07T10:20:30Z'));
    const about = render(
      <Provider store={store}>
        <Schedule />
      </Provider>
    );
    expect(about).toMatchSnapshot();
  });
});
