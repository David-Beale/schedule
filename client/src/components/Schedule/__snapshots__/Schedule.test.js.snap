// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schedule should render properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="outer-container"
      >
        <div
          class="container date"
        >
          <div
            class="arrow"
          >
            ⬅
          </div>
           
          <time
            class="sched-time"
            datetime="1554632430000"
          >
            Sun 7th Apr
          </time>
           
          <div
            class="arrow"
          >
            ➡
          </div>
        </div>
        <div
          class="container c1"
        >
          <div
            class="flex-container fc1"
          >
            <div
              class="grid-item1"
              id="margin0"
            >
              00:00
            </div>
            <div
              class="grid-item1"
              id="margin1"
            >
              01:00
            </div>
            <div
              class="grid-item1"
              id="margin2"
            >
              02:00
            </div>
            <div
              class="grid-item1"
              id="margin3"
            >
              03:00
            </div>
            <div
              class="grid-item1"
              id="margin4"
            >
              04:00
            </div>
            <div
              class="grid-item1"
              id="margin5"
            >
              05:00
            </div>
            <div
              class="grid-item1"
              id="margin6"
            >
              06:00
            </div>
            <div
              class="grid-item1"
              id="margin7"
            >
              07:00
            </div>
            <div
              class="grid-item1"
              id="margin8"
            >
              08:00
            </div>
            <div
              class="grid-item1"
              id="margin9"
            >
              09:00
            </div>
            <div
              class="grid-item1"
              id="margin10"
            >
              10:00
            </div>
            <div
              class="grid-item1"
              id="margin11"
            >
              11:00
            </div>
            <div
              class="grid-item1"
              id="margin12"
            >
              12:00
            </div>
            <div
              class="grid-item1"
              id="margin13"
            >
              13:00
            </div>
            <div
              class="grid-item1"
              id="margin14"
            >
              14:00
            </div>
            <div
              class="grid-item1"
              id="margin15"
            >
              15:00
            </div>
            <div
              class="grid-item1"
              id="margin16"
            >
              16:00
            </div>
            <div
              class="grid-item1"
              id="margin17"
            >
              17:00
            </div>
            <div
              class="grid-item1"
              id="margin18"
            >
              18:00
            </div>
            <div
              class="grid-item1"
              id="margin19"
            >
              19:00
            </div>
            <div
              class="grid-item1"
              id="margin20"
            >
              20:00
            </div>
            <div
              class="grid-item1"
              id="margin21"
            >
              21:00
            </div>
            <div
              class="grid-item1"
              id="margin22"
            >
              22:00
            </div>
            <div
              class="grid-item1 end"
            >
              23:00
            </div>
          </div>
        </div>
        <div
          class="container c2"
        >
          <div
            class="flex-container fc2"
            id="row0"
          >
            <div
              class="grid-item2"
              id="time0"
            />
            <div
              class="grid-item2"
              id="time1"
            />
            <div
              class="grid-item2"
              id="time2"
            />
            <div
              class="grid-item2"
              id="time3"
            />
            <div
              class="grid-item2"
              id="time4"
            />
            <div
              class="grid-item2"
              id="time5"
            />
            <div
              class="grid-item2"
              id="time6"
            />
            <div
              class="grid-item2"
              id="time7"
            />
            <div
              class="grid-item2"
              id="time8"
            />
            <div
              class="grid-item2"
              id="time9"
            />
            <div
              class="grid-item2"
              id="time10"
            />
            <div
              class="grid-item2"
              id="time11"
            >
              <div
                class="now"
                style="left: 33.33333333333333%; height: 400px;"
              />
            </div>
            <div
              class="grid-item2"
              id="time12"
            />
            <div
              class="grid-item2"
              id="time13"
            />
            <div
              class="grid-item2"
              id="time14"
            />
            <div
              class="grid-item2"
              id="time15"
            />
            <div
              class="grid-item2"
              id="time16"
            />
            <div
              class="grid-item2"
              id="time17"
            />
            <div
              class="grid-item2"
              id="time18"
            />
            <div
              class="grid-item2"
              id="time19"
            />
            <div
              class="grid-item2"
              id="time20"
            />
            <div
              class="grid-item2"
              id="time21"
            />
            <div
              class="grid-item2"
              id="time22"
            />
            <div
              class="grid-item2"
              id="time23"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="outer-container"
    >
      <div
        class="container date"
      >
        <div
          class="arrow"
        >
          ⬅
        </div>
         
        <time
          class="sched-time"
          datetime="1554632430000"
        >
          Sun 7th Apr
        </time>
         
        <div
          class="arrow"
        >
          ➡
        </div>
      </div>
      <div
        class="container c1"
      >
        <div
          class="flex-container fc1"
        >
          <div
            class="grid-item1"
            id="margin0"
          >
            00:00
          </div>
          <div
            class="grid-item1"
            id="margin1"
          >
            01:00
          </div>
          <div
            class="grid-item1"
            id="margin2"
          >
            02:00
          </div>
          <div
            class="grid-item1"
            id="margin3"
          >
            03:00
          </div>
          <div
            class="grid-item1"
            id="margin4"
          >
            04:00
          </div>
          <div
            class="grid-item1"
            id="margin5"
          >
            05:00
          </div>
          <div
            class="grid-item1"
            id="margin6"
          >
            06:00
          </div>
          <div
            class="grid-item1"
            id="margin7"
          >
            07:00
          </div>
          <div
            class="grid-item1"
            id="margin8"
          >
            08:00
          </div>
          <div
            class="grid-item1"
            id="margin9"
          >
            09:00
          </div>
          <div
            class="grid-item1"
            id="margin10"
          >
            10:00
          </div>
          <div
            class="grid-item1"
            id="margin11"
          >
            11:00
          </div>
          <div
            class="grid-item1"
            id="margin12"
          >
            12:00
          </div>
          <div
            class="grid-item1"
            id="margin13"
          >
            13:00
          </div>
          <div
            class="grid-item1"
            id="margin14"
          >
            14:00
          </div>
          <div
            class="grid-item1"
            id="margin15"
          >
            15:00
          </div>
          <div
            class="grid-item1"
            id="margin16"
          >
            16:00
          </div>
          <div
            class="grid-item1"
            id="margin17"
          >
            17:00
          </div>
          <div
            class="grid-item1"
            id="margin18"
          >
            18:00
          </div>
          <div
            class="grid-item1"
            id="margin19"
          >
            19:00
          </div>
          <div
            class="grid-item1"
            id="margin20"
          >
            20:00
          </div>
          <div
            class="grid-item1"
            id="margin21"
          >
            21:00
          </div>
          <div
            class="grid-item1"
            id="margin22"
          >
            22:00
          </div>
          <div
            class="grid-item1 end"
          >
            23:00
          </div>
        </div>
      </div>
      <div
        class="container c2"
      >
        <div
          class="flex-container fc2"
          id="row0"
        >
          <div
            class="grid-item2"
            id="time0"
          />
          <div
            class="grid-item2"
            id="time1"
          />
          <div
            class="grid-item2"
            id="time2"
          />
          <div
            class="grid-item2"
            id="time3"
          />
          <div
            class="grid-item2"
            id="time4"
          />
          <div
            class="grid-item2"
            id="time5"
          />
          <div
            class="grid-item2"
            id="time6"
          />
          <div
            class="grid-item2"
            id="time7"
          />
          <div
            class="grid-item2"
            id="time8"
          />
          <div
            class="grid-item2"
            id="time9"
          />
          <div
            class="grid-item2"
            id="time10"
          />
          <div
            class="grid-item2"
            id="time11"
          >
            <div
              class="now"
              style="left: 33.33333333333333%; height: 400px;"
            />
          </div>
          <div
            class="grid-item2"
            id="time12"
          />
          <div
            class="grid-item2"
            id="time13"
          />
          <div
            class="grid-item2"
            id="time14"
          />
          <div
            class="grid-item2"
            id="time15"
          />
          <div
            class="grid-item2"
            id="time16"
          />
          <div
            class="grid-item2"
            id="time17"
          />
          <div
            class="grid-item2"
            id="time18"
          />
          <div
            class="grid-item2"
            id="time19"
          />
          <div
            class="grid-item2"
            id="time20"
          />
          <div
            class="grid-item2"
            id="time21"
          />
          <div
            class="grid-item2"
            id="time22"
          />
          <div
            class="grid-item2"
            id="time23"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
